File: test/benchmark/r7rs/python/ack.py
#!/usr/bin/env python3
# R7RS Benchmark: ack (Python)
# Ackermann function - ack(3, 8) = 2045
import time
import sys

sys.setrecursionlimit(100000)


def ack(m, n):
    if m == 0:
        return n + 1
    if n == 0:
        return ack(m - 1, 1)
    return ack(m - 1, ack(m, n - 1))


def main():
    t0 = time.perf_counter_ns()
    result = ack(3, 8)
    t1 = time.perf_counter_ns()

    if result == 2045:
        print("ack: PASS")
    else:
        print(f"ack: FAIL result={result}")
    print(f"__TIMING__:{(t1 - t0) / 1e6:.3f}")


main()

File: test/benchmark/r7rs/c2mir/ack.c
/* Native C2MIR port of r7rs/ack2.ls. */
extern int printf(const char *, ...);

static int ack(int m, int n) {
    if (m == 0) return n + 1;
    if (n == 0) return ack(m - 1, 1);
    return ack(m - 1, ack(m, n - 1));
}

int main(void) {
    int result = ack(3, 8);
    printf(result == 2045 ? "ack: PASS\n" : "ack: FAIL result=%d\n", result);
    return result != 2045;
}
